crates/net/dns/src/lib.rs

```diff
@@ -356,29 +428,46 @@ impl<R: Resolver> DnsDiscoveryService<R> {
     }
 }
 
-/// A Stream events, mainly used for debugging
-impl<R: Resolver> Stream for DnsDiscoveryService<R> {
-    type Item = DnsDiscoveryEvent;
+/// Trait for converting an [`Enr`] into an update and sending the update. Trait is implemented
+/// for all supported update types.
+pub trait Update {
+    /// Tries to convert a resolved [`Enr`] into a [`DnsNodeRecordUpdate`]. Notifies all listeners
+    /// of update upon successful conversion.
+    fn on_resolved_enr(&mut self, enr: Enr<SecretKey>)
+        -> Result<(), DnsNodeRecordUpdateParseError>;
+}
+
+impl<R: Resolver> Update for DnsDiscoveryService<R> {
+    fn on_resolved_enr(
+        &mut self,
+        enr: Enr<SecretKey>,
+    ) -> Result<(), DnsNodeRecordUpdateParseError> {
+        self.queued_events.push_back(DnsDiscoveryEvent::Enr(enr.clone()));
+        let update = enr.try_into()?;
+        self.notify(update);
 
-    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
-        Poll::Ready(Some(ready!(self.get_mut().poll(cx))))
+        Ok(())
     }
 }
 
-/// The converted discovered [Enr] object
-#[derive(Debug, Clone, Eq, PartialEq)]
-pub struct DnsNodeRecordUpdate {
-    /// Discovered node and it's addresses
```

Comment:
isn't what we want the Enr as an additional field here?

Reply:
no why? the `NodeRecord` contains a subset of the data from the `Enr`. can still implement it if it's what you want.

Reply:
for purpose of learning, perf related: how I would do it is save a (not free) conversion like that until its needed, so the place that needs the other type, does the conversion. for example same rule goes for `Clone` trait - methods should rather return a reference to some type, and let the caller clone it if the caller needs ownership of the value, rather than the method cloning the type in its body.

Reply:
does an enum work here?

```rust
enum DnsRecordUpdate {
    V4NodeRecord(DnsNodeRecordUpdate), // could rename
    V5NodeRecord(Enr<SecretKey>)
}
```

Reply:
or just as an additional field, in the `DnsNodeRecordUpdate`
the point of the extracted NodeRecord type is that the work required for extracting the IP addresses and peerid is already done so we don't have to do this in the network loop, this also guarantees that the enr includes IP information.

Reply:
right, forgot that we already get `Enr`s from dnsdisc and convert earlier

Reply:
> IP information

ok this info make sense, had the `NodeRecordWithForkId` type in `emahane/discv5-support` pr for this, it's a more helpful type name than `DnsNodeRecordUpdate` to explain the constraints on the contained node record, i.e. (i) it has to convert to `NodeRecord`, requires IP, udp port, tcp port and ID, (ii) it additionally needs to have an "eth" kv-pair that decodes to `ForkId`. will have to make the key string configurable here to let this work with other networks. can make issue, keep scope minimal here.